core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -520,11 +520,15 @@ class TransactionStateManager(brokerId: Int,
                 // in this case directly return NOT_COORDINATOR to client and let it to re-discover the transaction coordinator
                 info(s"Updating $transactionalId's transaction state to $newMetadata with coordinator epoch $coordinatorEpoch for $transactionalId failed after the transaction message " +
                   s"has been appended to the log. The cached coordinator epoch has changed to ${epochAndMetadata.coordinatorEpoch}")
+                if (metadata.pendingState.isDefined)
```

Comment:
I think it is a legal state actually: upon immigration when we load txn metadata, if the metadata state is "prepareXX" it will try to complete the transaction by enqueueing the txn markers again and set its pending state.